### PR TITLE
[rayci] allow to restrict the list of concurrency group

### PR DESCRIPTION
**Pull request by can-anyscale**
- Allow to restrict the list of concurrency group; or allow everything if the restricted list is not provided
- Restrict the list of concurrency group for premerge and microcheck (empty list means that the concurrency group is not allowed)

Test:
- CI

### CODE REVIEW

**Comment by aslonnie** [comment]
> or we can also do the prefix or regex, but this approach will allow the trusted pipelines to have no restrictions

let's maybe do the prefix?

- the prefix allows user to add new concurrency groups without the need to change the config.
- regex is harder to reason about.. we do not seem to have use case for the complexity
- "allow the trusted pipelines to have no restrictions", this is fine I think; it is buildkite's default.

**Comment by can-anyscale** [comment]
sound good, i updated it to be a list of prefixes

File: raycicmd/command_step.go [hunk 1]
```diff
@@ -74,6 +74,20 @@ func (c *commandConverter) convert(id string, step map[string]any) (
 	if err != nil {
 		return nil, fmt.Errorf("map agent: %w", err)
 	}
+	concurrency_group, _ := stringInMap(step, "concurrency_group")
```
Comment (aslonnie) [inline raycicmd/command_step.go:77]: `concurrencyGroup`

go uses camal cases rather than underscores.

File: raycicmd/command_step.go [hunk 2]
```diff
@@ -74,6 +74,20 @@ func (c *commandConverter) convert(id string, step map[string]any) (
...
+	allowed_concurrency_group_prefixes := c.config.AllowConcurrencyGroupPrefixes
+	if concurrency_group != "" {
+		allowed_group := false
+		for _, group_prefix := range allowed_concurrency_group_prefixes {
+			if strings.HasPrefix(concurrency_group, group_prefix) {
+				allowed_group = true
+				break
+			}
+		}
```
Comment (aslonnie) [inline raycicmd/command_step.go:86]: have a helper function that returns boolean for this? and add some unit tests for the helper function?

File: raycicmd/command_step.go [hunk 3]
```diff
@@ -74,6 +74,20 @@ func (c *commandConverter) convert(id string, step map[string]any) (
 	if err != nil {
 		return nil, fmt.Errorf("map agent: %w", err)
 	}
+	concurrency_group, _ := stringInMap(step, "concurrency_group")
+	allowed_concurrency_group_prefixes := c.config.AllowConcurrencyGroupPrefixes
+	if concurrency_group != "" {
+		allowed_group := false
```
Comment (aslonnie) [inline raycicmd/command_step.go:80]: allow everything when the list is empty?

**Comment by can-anyscale** [comment]
a few changes to @aslonnie's comments
- changed the field name to `RestrictedConcurrencyGroupPrefixes`. If this field name doesn't exist then there will be no restriction on the group name; otherwise the concurrency group name needs to match with at least one prefix in the list (so empty list means no concurrency groups are allowed)
- created an util `stringInPrefix` and test cases
- update variables name to use camel case

File: raycicmd/config.go [hunk 4]
```diff
@@ -119,6 +119,12 @@ type config struct {
 	// Optional.
 	AllowTriggerStep bool `yaml:"allow_trigger_step"`
 
+	// RestrictedConcurrencyGroup is the list of concurrency group name prefixes that are
```
Comment (aslonnie) [inline raycicmd/config.go:122]: rename to `AllowedConcurrencyGroupPrefixes` ?

`Restricted` feels that it is not permitted to use..

File: raycicmd/converter_test.go [hunk 5]
```diff
@@ -80,6 +80,37 @@ func findInSlice(s []string, v string) bool {
...
+
+	step := map[string]any{
+		"label":             "say hello",
+		"key":               "key",
+		"command":           "echo hello",
+		"concurrency":       2,
+		"concurrency_group": "group",
+	}
+	_, err := c.convertStep("fakeid", step)
+	if err == nil {
```
Comment (aslonnie) [inline raycicmd/converter_test.go:109]: nit: `if _, err := ... ; err == nil {`

File: raycicmd/util.go [hunk 6]
```diff
@@ -13,6 +14,15 @@ func boolInMap(m map[string]any, key string) (bool, bool) {
 	return b, ok
 }
 
+func stringHasPrefix(prefixes []string, key string) bool {
```
Comment (aslonnie) [inline raycicmd/util.go:17]: nit: maybe put `key` first? as "string `key` has prefix `prefixes`? and rename `key` to `s`, as it is just a string, not a key?

the other ones here, like `stringInMap`, the string actually means "type" of the value. like `stringInMap` is checking if `m[key]` exists, and it is a string. quite a bit different from the one here.

File: raycicmd/command_step.go [hunk 7]
```diff
@@ -74,6 +74,14 @@ func (c *commandConverter) convert(id string, step map[string]any) (
 	if err != nil {
 		return nil, fmt.Errorf("map agent: %w", err)
 	}
+	concurrencyGroup, _ := stringInMap(step, "concurrency_group")
+	restrictedConcurrencyGroupPrefixes := c.config.RestrictedConcurrencyGroupPrefixes
+	if concurrencyGroup != "" && restrictedConcurrencyGroupPrefixes != nil {
```
Comment (aslonnie) [inline raycicmd/command_step.go:79]: you might want to check `len(restrictedConcurrencyGroupPrefixes) > 0` here. as it is quite hard to distinguish `nil` and `[]string{}` in yaml files.

File: raycicmd/command_step.go [hunk 8]
```diff
@@ -74,6 +74,14 @@ func (c *commandConverter) convert(id string, step map[string]any) (
...
+		allowed_group := stringHasPrefix(restrictedConcurrencyGroupPrefixes, concurrencyGroup)
```
Comment (aslonnie) [inline raycicmd/command_step.go:80]: camel case for `allowed_group` too? or maybe just write:

````
if g, ok := stringInMap(step, "concurrency_group"); ok {

  if prefixes := c.config.AllowedConcurrencyGroupPrefixes; len(prefixes) > 0 {
     if !stringsHasPrefix(prefixes, g) {
        return nil, ... // error here
     }
  }
}
````

File: raycicmd/command_step.go (hunk 7 above)
Comment (aslonnie) [reply raycicmd/command_step.go:79]: ok. that works too. could you add a comment here to make it clear that we treat nil and empty slice differently? in go, it is often treated the same in most parts of the common logic.

it is kind of like in python if one write `if list:` then it needs to be not None, also needs to be not an empty list.

File: raycicmd/config.go [hunk 9]
```diff
@@ -119,6 +119,12 @@ type config struct {
...
+	// allowed for the pipeline. If not set, all concurrency group names are allowed.
+	//
```
Comment (aslonnie) [inline raycicmd/config.go:124]: could you say "if set to `null` or left unset" in the comment here? it is actually possible to explicitly set to `null` here (I think it is `~` or `null`)

**Comment by can-anyscale** [comment]
@aslonnie's comments

**Review by aslonnie — APPROVED** [review]

**Pull request merged**